middle_end/flambda2/to_cmm/to_cmm_static.ml

```diff
@@ -23,34 +23,59 @@ module SC = Static_const
 module R = To_cmm_result
 module UK = C.Update_kind
 
-let static_value res v =
-  match (v : Field_of_static_block.t) with
-  | Symbol s -> C.symbol_address (R.symbol res s)
-  | Dynamically_computed _ -> C.cint 1n
-  | Tagged_immediate i ->
-    C.cint
-      (C.nativeint_of_targetint
-         (C.tag_targetint (Targetint_31_63.to_targetint i)))
+let static_field res field =
+  Simple.pattern_match'
+    (Simple.With_debuginfo.simple field)
+    ~var:(fun _var ~coercion:_ -> C.cint 1n)
+    ~symbol:(fun sym ~coercion:_ -> C.symbol_address (R.symbol res sym))
+    ~const:(fun cst ->
+      match Reg_width_const.descr cst with
+      | Tagged_immediate i ->
+        C.cint
+          (C.nativeint_of_targetint
+             (C.tag_targetint (Targetint_31_63.to_targetint i)))
+      | Naked_immediate i ->
+        C.cint (Targetint_31_63.to_int64 i |> Int64.to_nativeint)
+      | Naked_float32 f ->
+        C.cfloat32 (Numeric_types.Float32_by_bit_pattern.to_float f)
+      | Naked_float f ->
+        C.cfloat (Numeric_types.Float_by_bit_pattern.to_float f)
+      | Naked_int32 i -> C.cint32 (Nativeint.of_int32 i)
+      (* XCR nroberts: Could somebody check the below cases? They feel funny to
+         me.
+
+         mshinwell: they seem to be correct, what did you think is wrong? *)
+      | Naked_int64 i -> C.cint (Int64.to_nativeint i)
+      | Naked_nativeint i -> (
+        match Targetint_32_64.repr i with
+        | Int32 i -> C.cint (Nativeint.of_int32 i)
```

Comment:
This case is surprising considering that most of the code elsewhere assumes that we compile for a `64` bit target (and also that the host is `64` bit, else some conversions could lose some information).

Reply:
This has been removed now